### PR TITLE
Avoid loading all combinations in memory when only one is needed

### LINKED ISSUE #20928: Improve product combination / variants loading
#### Is your feature request related to a problem?

Some products can easily have thousands of combinations / variants. The current implementation is not optimal for such products and performance takes a big hit.
Improvements are already in progress for the admin product page and some improvements could also be implemented for the front end.

I currently have a product with over 2000 combinations. Changing between variants in the product page takes almost 3 seconds.

#### Describe the solution you'd like

At the moment, all combinations are queried from the database and loaded into memory in order to find the one with the correct id.  
Querying directly the desired id will improve performance.

PR to follow.

### PR DESCRIPTION
**Pull request by jbenezech**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Fixes slow loading of product page when the product has a lot of combination. Filters the desired combination in SQL instead of in-memory
| Type?         | improvement
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #20928
| How to test?  | composer run phpunit-legacy

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/20929)

### CODE REVIEW

**Comment by jbenezech** [comment]
This is in WIP for 2 reasons:

1. I cannot get the tests to run locally. Seems there are issues with loading fixture, but don't have more time to investigate. Letting travis do the job for me ;)

2. There is an existing method `getAttributeCombinationsById` (https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Product.php#L2719) which would be more logical to reuse but it doesn't seem to return exactly the same data as `getAttributesGroups` so I have no idea if that would break things down the line. I chose what seemed the safest option but definitely needs review by somebody with more insights into Prestashop internals.

**Comment by jbenezech** [comment]
Removed WIP status since tests have passed. My second comment still stands though

File: controllers/front/ProductController.php [hunk 1]
```diff
@@ -1100,17 +1100,13 @@ protected function getProductMinimalQuantity($product)
...
+        $combinations = $this->product->getAttributesGroups($this->context->language->id, $combinationId);
 
-                break;
-            }
+        if (empty($combinations)) {
+            return null;
         }
 
-        return $foundCombination;
+        return $combinations[0];
```
Comment (Progi1984) [inline controllers/front/ProductController.php]:
```suggestion
        return reset($combinations);
```

**Review by Progi1984 — COMMENTED** [review]
Good PR. But a small change

**Comment by jbenezech** [comment]
Thanks @Progi1984 . Fixed

**Review by Progi1984 — DISMISSED** [review]

**Comment by Progi1984** [comment]
@jbenezech Approved for me. Have you seen a performance improvement ? How ? Where ?

**Comment by jbenezech** [comment]
> @jbenezech Approved for me. Have you seen a performance improvement ? How ? Where ?

Manual checks only but time spent on the server side for my sample product (2560 combinations) went from 2.8s to 1.3s (on a dual CPU/4GB DO droplet). That `findProductCombinationById` being called twice.

The rest of the time is mostly spent here, for the exact same reason: https://github.com/PrestaShop/PrestaShop/blob/19dc654b76f7e45a465958d5308ef28c2dd5dac0/controllers/front/ProductController.php#L545

I'll open another issue when this is merged as the code in there is a bit more obscure and I don't really have time to spend on it right now.

**Review by atomiix — APPROVED** [review]

**Review by matthieu-rolland — DISMISSED** [review]

**Comment by ngodefroy** [comment]
@PrestaShop/prestashop-core-developers

File: controllers/front/ProductController.php [hunk 2]
```diff
@@ -1100,17 +1100,13 @@ protected function getProductMinimalQuantity($product)
...
+        $combinations = $this->product->getAttributesGroups($this->context->language->id, $combinationId);
 
-                break;
-            }
+        if (empty($combinations)) {
+            return null;
         }
 
-        return $foundCombination;
+        return reset($combinations);
```
Comment (PierreRambaud) [inline controllers/front/ProductController.php:1127]: The executeS method can return false. be sure it's an array :thinking:

Comment (matthieu-rolland) [reply controllers/front/ProductController.php:1127]: yep, that would trigger a warning

Comment (PierreRambaud) [reply controllers/front/ProductController.php:1127]: Yes, and Warning are catch in goes to exception :)

**Review by PierreRambaud — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Comment by jbenezech** [comment]
Apologies for the delay.  
Any idea where this Travis error comes from ?  
`Could not scan for classes inside "modules/ps_themecusto//ps_themecusto.php" which does not appear to be a file nor a folder`

**Comment by Progi1984** [comment]
@jbenezech I restared all failing jobs.

**Review by atomiix — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by SD1982** [comment]
@PrestaShop/prestashop-core-developers can we let you handle this one ?

**Comment by PierreRambaud** [comment]
> @PrestaShop/prestashop-core-developers can we let you handle this one ?

Of course!

**Comment by atomiix** [comment]
@jbenezech I just did some tests with and without your PR, there is no significant memory consumption improvement because the function is still called with `$id_product_attribute = null` at some point to show every possible combination on the FO product page and then as the request is cached, if you call it again, it will reuse the cached results.
I'm still validating this PR as it avoid an (potentially big) array iteration.

It's QA ✔️

**Comment by PierreRambaud** [comment]
Thank you @jbenezech

**Pull request merged**